pkg/river/json_walker.go

```diff
@@ -0,0 +1,409 @@
+	// If this is a pointer then its an interface{}nil and return
+	if inType.Kind() == reflect.Pointer {
+		return nil, nil
+	}
+	// If its a struct that is entirely default values return nil
+	if inType.Kind() == reflect.Struct && inValue.IsZero() {
+		return nil, nil
+	}
+
+	if f != nil && f.IsBlock() {
+		return convertBlock(in, f)
+	}
+
+	// Capsules would be seen as a struct below but we want to treat them special.
+	if value.RiverType(inType) == value.TypeCapsule {
+		return convertValue(in)
+	}
+
+	if inType.Kind() == reflect.Array || inType.Kind() == reflect.Slice {
+		return convertArray(inType, inValue, f)
+	} else if inType.Kind() == reflect.Struct {
+		if f != nil && f.IsAttr() {
+			v, err := convertAttribute(in, inType, f)
+			if err != nil {
+				return nil, err
+			}
+			return v, nil
+		} else if !rivertags.HasRiverTags(inType) {
+			// This is used when the caller is converting a struct that is technically a struct, but should be
+			// treated like a string. Regex, time.time, time.duration etc. If it doesn't have river tags then we
+			// get the string representation.
+			v, err := convertValue(in)
+			if err != nil {
+				return nil, err
+			}
+			return v, nil
+		} else {
+			return convertStruct(in, inValue, f)
+		}
+	} else if inType.Kind() == reflect.Map {
+		return convertMap(inValue)
+	}
+	if f != nil {
+		return convertAttribute(in, inType, f)
+	}
+	return convertValue(in)
+}
+
+func convertAttribute(in interface{}, t reflect.Type, f *rivertags.Field) (*Field, error) {
+	if !f.IsAttr() {
+		return nil, fmt.Errorf("convertAttribute requires a field that is an attribute got %T", in)
+	}
+	nf := &Field{
+		Name: strings.Join(f.Name, "."),
+		Type: attr,
+	}
+
+	// If this is a simple attribute then grab the value directly.
+	if isValue(t) {
+		v, err := convertValue(in)
+		if err != nil {
+			return nil, err
+		}
+		nf.Value = v
+	} else {
+		v, err := convertThing(in, nil)
+		if err != nil {
+			return nil, err
+		}
+		nf.Value = v
+	}
+	if nf.Value == nil || reflect.ValueOf(nf.Value).IsZero() {
+		return nil, nil
+	}
+	return nf, nil
+}
+
+func convertArray(inType reflect.Type, inValue reflect.Value, f *rivertags.Field) (*Field, error) {
+	if inType.Kind() != reflect.Array && inType.Kind() != reflect.Slice {
+		return nil, fmt.Errorf("convertArray requires a field that is an slice/array got %T", inValue.Interface())
+	}
+	nf := &Field{}
+	nf.Type = "array"
+
+	blocks := make([]interface{}, 0)
+	for i := 0; i < inValue.Len(); i++ {
+		arrEle := inValue.Index(i).Interface()
+		found, err := convertThing(arrEle, f)
+		if err != nil {
+			return nil, err
+		}
+		if found != nil {
+			blocks = append(blocks, found)
+		}
+	}
+
+	nf.Value = blocks
+	return nf, nil
+}
+
+func isValue(t reflect.Type) bool {
+	rt := value.RiverType(t)
+	switch rt {
+	case value.TypeNull, value.TypeNumber, value.TypeString, value.TypeBool, value.TypeFunction, value.TypeCapsule:
+		return true
+	}
+	return false
+}
```

Comment:
Unless I'm missing something, this always returns true

Reply:
Oh I see, you're not considering arrays or objects to be values, but they are. I think this might need a more clear name